From 8369abf1cd64603827ac994668fb6329979b6b16 Mon Sep 17 00:00:00 2001
From: Stefan Marr <git@stefan-marr.de>
Date: Sun, 7 Feb 2016 22:21:06 +0100
Subject: [PATCH 1/3] Added helpers for on-demand instrumentation

This is useful for dynamic languages where execution leads to node
replacement, which requires new instrumentation.

This change is a *hack*, and not supposed to be merged directly.

Handled special cases:

- unwrap wrapper nodes when inserting new wrappers
- when there is already a wrapper node, just invalidate it

Signed-off-by: Stefan Marr <git@stefan-marr.de>
---
 .../instrumentation/InstrumentationHandler.java    | 23 +++++++++++++++++++++-
 1 file changed, 22 insertions(+), 1 deletion(-)

diff --git a/truffle/src/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/InstrumentationHandler.java b/truffle/src/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/InstrumentationHandler.java
index 0a8d5a62f7..f068495dd9 100644
--- a/truffle/src/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/InstrumentationHandler.java
+++ b/truffle/src/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/InstrumentationHandler.java
@@ -74,11 +74,13 @@ import com.oracle.truffle.api.source.SourceSection;
  * Central coordinator class for the Truffle instrumentation framework. Allocated once per
  * {@linkplain com.oracle.truffle.api.vm.PolyglotEngine engine}.
  */
-final class InstrumentationHandler {
+public final class InstrumentationHandler {
 
     /* Enable trace output to stdout. */
     static final boolean TRACE = Boolean.getBoolean("truffle.instrumentation.trace");
 
+    private static InstrumentationHandler globalHandler;
+
     private final Object sourceVM;
 
     /*
@@ -137,6 +139,24 @@ final class InstrumentationHandler {
         this.err = err;
         this.in = in;
         this.engineInstrumenter = new EngineInstrumenter();
+        globalHandler = this;
+    }
+
+    public static void insertInstrumentationWrapper(Node instrumentableNode) {
+        insertInstrumentationWrapper(instrumentableNode, instrumentableNode.getSourceSection());
+    }
+
+    public static void insertInstrumentationWrapper(Node instrumentableNode, SourceSection sourceSection) {
+        assert globalHandler != null : "InstrumentationHandler not yet initialized";
+
+        Node node;
+        if (instrumentableNode instanceof WrapperNode) {
+            node = ((WrapperNode) instrumentableNode).getDelegateNode();
+            invalidateWrapperImpl((WrapperNode) instrumentableNode, node);
+        } else {
+            node = instrumentableNode;
+            globalHandler.insertWrapper(node, sourceSection);
+        }
     }
 
     Object getSourceVM() {
@@ -824,6 +844,7 @@ final class InstrumentationHandler {
     private void insertWrapperImpl(Node originalNode, SourceSection sourceSection) {
         Node node = originalNode;
         Node parent = node.getParent();
+        assert !(instrumentableNode instanceof WrapperNode);
         if (parent instanceof com.oracle.truffle.api.instrumentation.InstrumentableFactory.WrapperNode) {
             // already wrapped, need to invalidate the wrapper something changed
             invalidateWrapperImpl((com.oracle.truffle.api.instrumentation.InstrumentableFactory.WrapperNode) parent, node);
-- 
2.15.1 (Apple Git-101)


From 4ddced6472b3eb061906b55841a993855b662764 Mon Sep 17 00:00:00 2001
From: Christian Humer <christian.humer@oracle.com>
Date: Wed, 17 Feb 2016 13:41:48 +0100
Subject: [PATCH 2/3] Add query methods to EventContext to find parent or child
 nodes

- implement EventContext#findParentEventNode and EventContext#findChildEventNodes
  - also add a more specific method that only returns event node from
    direct parent
- updated EventNode[Factory] reference to ExecutionEventNode[Factory],
  these types were renamed.
- work around javac failing to compiler: move classes to top level
- change name to be more descriptive since field was moved to outer class

Signed-off-by: Stefan Marr <git@stefan-marr.de>
Co-authored-by: Stefan Marr <git@stefan-marr.de>
---
 .../instrumentation/test/InstrumentationTest.java  | 117 +++++++++++++++++++++
 .../truffle/api/instrumentation/EventContext.java  |  78 ++++++++++++++
 .../truffle/api/instrumentation/ProbeNode.java     |  20 +++-
 3 files changed, 214 insertions(+), 1 deletion(-)

diff --git a/truffle/src/com.oracle.truffle.api.instrumentation.test/src/com/oracle/truffle/api/instrumentation/test/InstrumentationTest.java b/truffle/src/com.oracle.truffle.api.instrumentation.test/src/com/oracle/truffle/api/instrumentation/test/InstrumentationTest.java
index 474cea4112..0673a8cde3 100644
--- a/truffle/src/com.oracle.truffle.api.instrumentation.test/src/com/oracle/truffle/api/instrumentation/test/InstrumentationTest.java
+++ b/truffle/src/com.oracle.truffle.api.instrumentation.test/src/com/oracle/truffle/api/instrumentation/test/InstrumentationTest.java
@@ -40,6 +40,7 @@ import java.io.PipedOutputStream;
 import java.io.Reader;
 import java.nio.CharBuffer;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
@@ -1180,6 +1181,122 @@ public class InstrumentationTest extends AbstractInstrumentationTest {
         }
     }
 
+    /*
+     * Test that parsing and executing foreign languages with context work.
+     */
+    @Test
+    public void testTestFindParentEventNode1() throws IOException {
+        InstrumentationTest.findParentNodes = new ArrayList<>();
+        List<FindParentEventNode> nodes = InstrumentationTest.findParentNodes;
+
+        assureEnabled(engine.getInstruments().get("testTestFindParentEventNode1"));
+        run("STATEMENT(EXPRESSION,EXPRESSION(STATEMENT))");
+
+        // assert no assertions during execution
+        Assert.assertTrue("Errors during execution: " + getErr(), getErr().isEmpty());
+
+        Assert.assertEquals(4, nodes.size());
+
+        FindParentEventNode node0 = nodes.get(0);
+        FindParentEventNode node1 = nodes.get(1);
+        FindParentEventNode node2 = nodes.get(2);
+        FindParentEventNode node3 = nodes.get(3);
+
+        Assert.assertNull(node0.parentNode);
+        Assert.assertSame(node1.parentNode, node0);
+        Assert.assertSame(node2.parentNode, node0);
+        Assert.assertSame(node3.parentNode, node2);
+
+        Assert.assertTrue(node0.beforeChildren.get(0).isEmpty());
+        Assert.assertTrue(node1.beforeChildren.get(0).isEmpty());
+        Assert.assertTrue(node2.beforeChildren.get(0).isEmpty());
+        Assert.assertTrue(node3.beforeChildren.get(0).isEmpty());
+
+        Assert.assertEquals(Arrays.asList(node1, node2), node0.afterChildren.get(0));
+        Assert.assertEquals(Arrays.asList(), node1.afterChildren.get(0));
+        Assert.assertEquals(Arrays.asList(node3), node2.afterChildren.get(0));
+        Assert.assertEquals(Arrays.asList(), node3.afterChildren.get(0));
+
+        run("STATEMENT(EXPRESSION,EXPRESSION(STATEMENT))");
+        // assert no assertions during execution
+        Assert.assertTrue("Errors during execution: " + getErr(), getErr().isEmpty());
+
+        Assert.assertEquals(4, nodes.size());
+
+        node0 = nodes.get(0);
+        node1 = nodes.get(1);
+        node2 = nodes.get(2);
+        node3 = nodes.get(3);
+
+        Assert.assertNull(node0.parentNode);
+        Assert.assertSame(node1.parentNode, node0);
+        Assert.assertSame(node2.parentNode, node0);
+        Assert.assertSame(node3.parentNode, node2);
+
+        Assert.assertEquals(Arrays.asList(node1, node2), node0.beforeChildren.get(1));
+        Assert.assertEquals(Arrays.asList(), node1.beforeChildren.get(1));
+        Assert.assertEquals(Arrays.asList(node3), node2.beforeChildren.get(1));
+        Assert.assertEquals(Arrays.asList(), node3.beforeChildren.get(1));
+
+        Assert.assertEquals(Arrays.asList(node1, node2), node0.afterChildren.get(1));
+        Assert.assertEquals(Arrays.asList(), node1.afterChildren.get(1));
+        Assert.assertEquals(Arrays.asList(node3), node2.afterChildren.get(1));
+        Assert.assertEquals(Arrays.asList(), node3.afterChildren.get(1));
+
+    }
+
+    @Registration(id = "testTestFindParentEventNode1", services = Object.class)
+    public static class TestFindParentEventNode1 extends TruffleInstrument {
+
+        static int parentNodesFound = 0;
+
+        @Override
+        protected void onCreate(final Env env) {
+            env.getInstrumenter().attachFactory(SourceSectionFilter.newBuilder().tagIs(InstrumentationTestLanguage.STATEMENT, InstrumentationTestLanguage.EXPRESSION).build(),
+                            new FindParentExecutionEventNodeFactory());
+        }
+    }
+
+    static class FindParentExecutionEventNodeFactory implements ExecutionEventNodeFactory {
+        public ExecutionEventNode create(final EventContext context) {
+            return new FindParentEventNode(context, this);
+        }
+    }
+
+    static List<FindParentEventNode> findParentNodes = new ArrayList<>();
+
+    static class FindParentEventNode extends ExecutionEventNode {
+
+        private final EventContext context;
+        private final ExecutionEventNodeFactory factory;
+
+        FindParentEventNode(EventContext context, ExecutionEventNodeFactory factory) {
+            this.context = context;
+            this.factory = factory;
+            findParentNodes.add(this);
+        }
+
+        ExecutionEventNode parentNode;
+        List<List<ExecutionEventNode>> beforeChildren = new ArrayList<>();
+        List<List<ExecutionEventNode>> afterChildren = new ArrayList<>();
+
+        @Override
+        protected void onEnter(VirtualFrame frame) {
+            ExecutionEventNode parent = context.findParentEventNode(factory);
+            if (this.parentNode != null) {
+                Assert.assertSame(parent, parentNode);
+            }
+            this.parentNode = parent;
+            this.beforeChildren.add(context.findChildEventNodes(factory));
+        }
+
+        @Override
+        protected void onReturnValue(VirtualFrame frame, Object result) {
+            Assert.assertSame(parentNode, context.findParentEventNode(factory));
+            this.afterChildren.add(context.findChildEventNodes(factory));
+        }
+    }
+
     /*
      * Tests for debugger or any other clients that cancel execution while halted
      */
diff --git a/truffle/src/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/EventContext.java b/truffle/src/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/EventContext.java
index a1c3f541e0..f4e5e2e12a 100644
--- a/truffle/src/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/EventContext.java
+++ b/truffle/src/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/EventContext.java
@@ -25,12 +25,16 @@
 package com.oracle.truffle.api.instrumentation;
 
 import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
 import java.util.Objects;
 
 import com.oracle.truffle.api.CallTarget;
 import com.oracle.truffle.api.CompilerAsserts;
 import com.oracle.truffle.api.CompilerDirectives;
 import com.oracle.truffle.api.CompilerDirectives.CompilationFinal;
+import com.oracle.truffle.api.CompilerDirectives.TruffleBoundary;
 import com.oracle.truffle.api.TruffleLanguage.Env;
 import com.oracle.truffle.api.frame.VirtualFrame;
 import com.oracle.truffle.api.instrumentation.InstrumentationHandler.AccessorInstrumentHandler;
@@ -298,6 +302,80 @@ public final class EventContext {
         return new UnwindException(info, unwindBinding);
     }
 
+    /**
+     * Returns the first found parent {@link ExecutionEventNode event node} created from a given
+     * {@link ExecutionEventNodeFactory factory}. If multiple
+     * {@link Instrumenter#attachFactory(SourceSectionFilter, ExecutionEventNodeFactory) bindings}
+     * were created with a single {@link ExecutionEventNodeFactory factory} instance then the first
+     * ExecutionEventNode which is found is returned in the order of event binding attachment.
+     *
+     * @param factory a event node factory for which to return the first event node
+     * @return the first event node found in the order of event binding attachment
+     * @since 0.13
+     */
+    @TruffleBoundary
+    public ExecutionEventNode findParentEventNode(final ExecutionEventNodeFactory factory) {
+        Node parent = getInstrumentedNode().getParent();
+        while ((parent = parent.getParent()) != null) {
+            ExecutionEventNode eventNode = findEventNode(factory, parent);
+            if (eventNode != null) {
+                return eventNode;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * @return an event node from the direct parent, or null.
+     * @since 0.13
+     */
+    @TruffleBoundary
+    public ExecutionEventNode findDirectParentEventNode(final ExecutionEventNodeFactory factory) {
+        Node parent = getInstrumentedNode().getParent();
+
+        assert parent instanceof WrapperNode;  // this is the wrapper of the current node
+        parent = parent.getParent();           // this is the parent node
+        parent = parent.getParent();           // this is the wrapper of the parent node
+        return findEventNode(factory, parent);
+    }
+
+    /**
+     * Returns all first-level child event nodes created from a given
+     * {@link ExecutionEventNodeFactory factory}.
+     *
+     * @param factory an event node factory for which to return all first-level children
+     * @return all first-level children that were created from a given factory
+     * @since 0.13
+     */
+    @TruffleBoundary
+    public List<ExecutionEventNode> findChildEventNodes(final ExecutionEventNodeFactory factory) {
+        final List<ExecutionEventNode> eventNodes = new ArrayList<>();
+        Node instrumentedNode = getInstrumentedNode();
+        // TODO ideally one could use a NodeListener instead of the recursive algortihm.
+        // Unfortunately returning false in NodeVisitor#visit does not continue traversing all
+        // parents children but stops visitation completely. Bug!?
+        collectEventNodes(eventNodes, factory, instrumentedNode);
+        return Collections.unmodifiableList(eventNodes);
+    }
+
+    private void collectEventNodes(List<ExecutionEventNode> eventNodes, ExecutionEventNodeFactory factory, Node node) {
+        for (Node child : node.getChildren()) {
+            ExecutionEventNode eventNode = findEventNode(factory, child);
+            if (eventNode != null) {
+                eventNodes.add(eventNode);
+            } else if (child != null) {
+                collectEventNodes(eventNodes, factory, child);
+            }
+        }
+    }
+
+    private static ExecutionEventNode findEventNode(ExecutionEventNodeFactory factory, Node node) {
+        if (node instanceof WrapperNode) {
+            return ((WrapperNode) node).getProbeNode().findEventNode(factory);
+        }
+        return null;
+    }
+
     /*
      * TODO (chumer) a way to parse code in the current language and return something like a node
      * that is directly embeddable into the AST as a @Child.
diff --git a/truffle/src/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/ProbeNode.java b/truffle/src/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/ProbeNode.java
index c9b373b266..80b847bea2 100644
--- a/truffle/src/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/ProbeNode.java
+++ b/truffle/src/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/ProbeNode.java
@@ -323,6 +323,24 @@ public final class ProbeNode extends Node {
         }
     }
 
+    ExecutionEventNode findEventNode(final ExecutionEventNodeFactory factory) {
+        if (version != null && version.isValid() && chain != null) {
+            return findEventNodeInChain(factory);
+        }
+        return null;
+    }
+
+    private ExecutionEventNode findEventNodeInChain(ExecutionEventNodeFactory factory) {
+        EventChainNode currentChain = this.chain;
+        while (currentChain != null) {
+            if (currentChain.binding.getElement() == factory) {
+                return ((EventProviderChainNode) currentChain).eventNode;
+            }
+            currentChain = currentChain.next;
+        }
+        return null;
+    }
+
     EventChainNode lazyUpdate(VirtualFrame frame) {
         Assumption localVersion = this.version;
         if (localVersion == null || !localVersion.isValid()) {
@@ -494,7 +512,7 @@ public final class ProbeNode extends Node {
         try {
             eventNode = ((ExecutionEventNodeFactory) element).create(context);
             if (eventNode.getParent() != null) {
-                throw new IllegalStateException(String.format("Returned EventNode %s was already adopted by another AST.", eventNode));
+                throw new IllegalStateException(String.format("Returned ExecutionEventNode %s was already adopted by another AST.", eventNode));
             }
         } catch (Throwable t) {
             if (binding.isLanguageBinding()) {
-- 
2.15.1 (Apple Git-101)


From 949452f10e9ab4a4360c9336ad3e515ec7bf1e8c Mon Sep 17 00:00:00 2001
From: Stefan Marr <git@stefan-marr.de>
Date: Tue, 5 Apr 2016 23:26:14 +0200
Subject: [PATCH 3/3] Added isTaggedWith(Node,Class<?>) to Instrumenter

Signed-off-by: Stefan Marr <git@stefan-marr.de>
---
 .../api/instrumentation/InstrumentationHandler.java        | 14 ++++++++++++++
 .../oracle/truffle/api/instrumentation/Instrumenter.java   |  1 +
 2 files changed, 15 insertions(+)

diff --git a/truffle/src/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/InstrumentationHandler.java b/truffle/src/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/InstrumentationHandler.java
index f068495dd9..af88aa7bd4 100644
--- a/truffle/src/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/InstrumentationHandler.java
+++ b/truffle/src/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/InstrumentationHandler.java
@@ -1465,6 +1465,11 @@ public final class InstrumentationHandler {
             return queryTagsImpl(node, null);
         }
 
+        @Override
+        public boolean isTaggedWith(Node node, Class<?> tag) {
+            return AccessorInstrumentHandler.nodesAccess().isTaggedWith(node, tag);
+        }
+
         @Override
         void verifyFilter(SourceSectionFilter filter) {
         }
@@ -1606,6 +1611,10 @@ public final class InstrumentationHandler {
             throw new UnsupportedOperationException("Not supported in engine instrumenter.");
         }
 
+        @Override
+        public boolean isTaggedWith(Node node, Class<?> tag) {
+            return AccessorInstrumentHandler.nodesAccess().isTaggedWith(node, tag);
+        }
     }
 
     /**
@@ -1646,6 +1655,11 @@ public final class InstrumentationHandler {
             return queryTagsImpl(node, languageInfo);
         }
 
+        @Override
+        public boolean isTaggedWith(Node node, Class<?> tag) {
+            return AccessorInstrumentHandler.nodesAccess().isTaggedWith(node, tag);
+        }
+
         @Override
         void verifyFilter(SourceSectionFilter filter) {
             Set<Class<?>> providedTags = getProvidedTags(languageInfo);
diff --git a/truffle/src/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/Instrumenter.java b/truffle/src/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/Instrumenter.java
index 0fb4ed8a98..33d8ed8d76 100644
--- a/truffle/src/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/Instrumenter.java
+++ b/truffle/src/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/Instrumenter.java
@@ -349,4 +349,5 @@ public abstract class Instrumenter {
      */
     public abstract Set<Class<?>> queryTags(Node node);
 
+    public abstract boolean isTaggedWith(Node node, Class<?> tag);
 }
-- 
2.15.1 (Apple Git-101)

